### ISSUE TITLE
Add ZTransducer arrow operations

### ISSUE THREAD

**Issue by iravid** [zio/zio#3495]
ZTransducer is an effectful chunk processing function, and thus forms an arrow. We should add the following combinators on it:
- [x] zipping: `(ZTransducer[A, B], ZTransducer[A, C]) => ZTransducer[A, (B, C)]`
- [ ] ~racing: `(ZTransducer[A, B], ZTransducer[A, C]) => ZTransducer[A, Either[B, C]]`~ looks problematic to implement currently
- [ ] both: `(ZTransducer[A, C], ZTransducer[B, D]) => ZTransducer[(A, C), (B, D)]`
- [ ] choose: `(ZTransducer[A, C], ZTransducer[B, D]) => ZTransducer[Either[A, C], Either[B, D]]`

Part of this issue would be to choose the appropriate names and see whether there are more generalized versions of these (e.g. for choose, a version where the types can unify through subtyping without `Either`)

We should also check for similarities between these implementations and `ZStream#zip/merge`, and figure out if we should move the `ZStream` implementations here.

**Comment by simpadjo** [comment]
In `race`, do we remove the loser from the competition?

Consider:
`t1` output:
 ---B1-----------B2----------
`t2` output:
------------C1---------C2-----

should `t1.race(t2)` emit  
----B1----C1---B2--C2-----

or
----B1-----------B2----------
?

**Comment by simpadjo** [comment]
No one is hungry of good first issues so let me take it. I'll start with `zip`

**Comment by iravid** [comment]
@simpadjo Yes, we should remove the loser from the competition. But! We must not interrupt the loser. That's currently against the ZStream protocol. We should just wait for it to finish, continuing to feed data to the winner in the meantime and making sure to discard its result when/if it does finish.

**Comment by fagossa** [comment]
I would like to take `ZTransducer#both`

**Comment by simpadjo** [comment]
`zip` is done, picking `race` now

**Comment by simpadjo** [comment]
@iravid If implemented naively, result of `race` depends on chunking strategy.
If both sides emitted a value when being fed a huge chunk, there is no way to determine who was the fastest. `choose` will suffer from the same issue.

We can either relax guarantees or feed upstream chunks with elements one-by-one.
What do you think?

**Comment by simpadjo** [comment]
Also taking `choose` since it can be easily implemented in terms of `race`

**Comment by simpadjo** [comment]
@fagossa Hey! Seems that `both` could be implemented using `zip` (like `choose` using `race`).
Here is the PR with `zip` : https://github.com/zio/zio/pull/3551

**Comment by fagossa** [comment]
@simpadjo Great, I'll take a look

**Comment by simpadjo** [comment]
Just sending a healthcheck that I didn't abandon `choose`

**Comment by iravid** [comment]
Thank you @simpadjo

**Comment by kjanosz** [comment]
I would like to pick ZTransducer#both, since I see no work done there yet.

**Comment by iravid** [comment]
@kjanosz ZTransducer will be removed in the upcoming version of ZIO Streams, so I would suggest not spending time on that 🙏🏻 Thank you for pitching in!
